cmd/security-agent/app/compliance.go

```diff
@@ -64,16 +77,118 @@ func eventRun(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("failed to set up compliance log reporter: %w", err)
 	}
 
-	eventArgs.event.Data = event.Data{}
+	eventData := event.Data{}
 	for _, d := range eventArgs.data {
 		kv := strings.SplitN(d, ":", 2)
 		if len(kv) != 2 {
 			continue
 		}
-		eventArgs.event.Data[kv[0]] = kv[1]
+		eventData[kv[0]] = kv[1]
 	}
+	eventArgs.event.Data = eventData
 
 	reporter.Report(&eventArgs.event)
 
 	return nil
 }
+
+func newComplianceReporter(stopper restart.Stopper, sourceName, sourceType string) (event.Reporter, error) {
+	httpConnectivity := config.HTTPConnectivityFailure
+	if endpoints, err := config.BuildHTTPEndpoints(); err == nil {
+		httpConnectivity = http.CheckConnectivity(endpoints.Main)
+	}
+
+	endpoints, err := config.BuildEndpoints(httpConnectivity)
+	if err != nil {
+		return nil, log.Errorf("Invalid endpoints: %v", err)
+	}
+
+	destinationsCtx := client.NewDestinationsContext()
+	destinationsCtx.Start()
+	stopper.Add(destinationsCtx)
+
+	health := health.RegisterLiveness("security-agent")
+
+	// setup the auditor
+	auditor := auditor.New(coreconfig.Datadog.GetString("compliance_config.run_path"), health)
+	auditor.Start()
+	stopper.Add(auditor)
+
+	// setup the pipeline provider that provides pairs of processor and sender
+	pipelineProvider := pipeline.NewProvider(config.NumberOfPipelines, auditor, nil, endpoints, destinationsCtx)
+	pipelineProvider.Start()
+	stopper.Add(pipelineProvider)
+
+	logSource := config.NewLogSource(
+		sourceName,
+		&config.LogsConfig{
+			Type:    sourceType,
+			Service: sourceName,
+			Source:  sourceName,
+		},
+	)
+	return event.NewReporter(logSource, pipelineProvider.NextPipelineChan()), nil
+}
+
+func startCompliance(stopper restart.Stopper) error {
+	enabled := coreconfig.Datadog.GetBool("compliance_config.enabled")
+	if !enabled {
+		return nil
+	}
+
+	reporter, err := newComplianceReporter(stopper, "compliance-agent", "compliance")
+	if err != nil {
+		return err
+	}
+
+	runner := runner.NewRunner()
+	stopper.Add(runner)
+
+	scheduler := scheduler.NewScheduler(runner.GetChan())
+	runner.SetScheduler(scheduler)
+
+	checkInterval := coreconfig.Datadog.GetDuration("compliance_config.check_interval")
+	configDir := coreconfig.Datadog.GetString("compliance_config.dir")
+
+	hostname, err := util.GetHostname()
```

Comment:
The gRPC endpoint here is available and ready for use. No authentication is necessary to use this endpoint and is available at `pb.Agent.GetHostname` [here](https://github.com/DataDog/datadog-agent/blob/master/cmd/agent/api/pb/api.proto#L10-L14). Also available at https://localhost:5001/v1/grpc/host, via the grpc-gateway, but the protobuf/gRPC way would be the preferred approach.

Let me know if any help is needed with this. Something to keep in mind:
1) be resilient to endpoint failures, although we can set service dependencies on systemd it might not be enough to guarantee the API is up and fully ready by the time the security agent queries for the canonical hostname, so I recommend adding some retry mechanism.
2) if we fail to collect the hostname via the API, then we can fallback to the legacy way of grabbing the hostname via a shellout.

Reply:
Makes sense.
The security-agent is built using the `docker` and `kubelet` build tags so most hostname provider will work. If we later choose to forward the logs to the agent, we would not need to retrieve the hostname at all. If this is ok for you @olivielpeau and @truthbk, I suggest we left it as it until we decide whether we keep using a dedicated log client or forward the logs to the agent.

Reply:
ok, I quickly checked the implementation of `util.GetHostname` and at the moment the `docker` and `kubelet` build tags seem to be the only ones that can affect it.

Your plan sounds good to me, but please make sure this is tracked on your end in case additional build tags in the future affect the hostname resolution logic. Hostname inconsistencies can have billing implications. Also, can you add a `FIXME` comment inline?

Reply:
Ok. Also added a FIXME